### DISCUSSION TITLE
Building a new PrePARE

### DISCUSSION THREAD

**Discussion by mauzey1** [PCMDI/cmor#690] [Ideas]
Continuing a discussion from https://github.com/PCMDI/cmor/issues/685#issuecomment-1403872151, I wanted to discuss the idea of creating a new PrePARE written entirely in Python using [xarray](https://github.com/pydata/xarray).

The idea is to create a validation tool that is similar to PrePARE but relies more on the information provided in CMIP tables rather than hard-coded into the program.  Rather than using the CMIP6-centric CV library built into CMOR, the new PrePARE should use Python-based tools that perform similar functions to CMOR.  Python provides built-in tools for reading JSON files making ingestion of the CMIP tables easy.  The list and dictionary structures of Python provide a much easier way of searching through these tables than what we have in our C code.

The xarray package provides the capability to open [NetCDF files](https://docs.xarray.dev/en/stable/user-guide/io.html?highlight=netcdf#netcdf) as labeled, multidimensional arrays similar to pandas DataFrames.  xarray is also being used by the [xCDAT](https://github.com/xCDAT/xcdat) team to create the successor to CDAT so the new PrePARE could be part of a software ecosystem centered around xarray within the LLNL climate program.

This tool should be flexible enough to perform validation of datasets for CMIP6, obs4MIPs, input4MIPS, CORDEX, and any other tables derived from [cmip6-cmor-tables](https://github.com/PCMDI/cmip6-cmor-tables).  It should also be able to handle the structure of the future [mip-cmor-tables](https://github.com/PCMDI/mip-cmor-tables) being prototyped.

These are my current thoughts on this.  Other thoughts and suggestions are welcomed.

**Comment by larsbuntemeyer** [comment]
That sounds reasonably to me! Just wanted to quickly mention [cf-xarray](https://cf-xarray.readthedocs.io/en/latest/examples/introduction.html) here, which is an accessor to xarray that understands CF conventions reasonably well. That might come in very useful for such a tool!
